lib/model/binding.dart

```diff
@@ -164,11 +262,68 @@ class LiveZulipBinding extends ZulipBinding {
   /// Initialize the binding if necessary, and ensure it is a [LiveZulipBinding].
   static LiveZulipBinding ensureInitialized() {
     if (ZulipBinding._instance == null) {
-      LiveZulipBinding();
+      final binding = LiveZulipBinding();
+      binding._deviceInfo = binding._prefetchDeviceInfo();
+      binding._packageInfo = binding._prefetchPackageInfo();
     }
     return ZulipBinding.instance as LiveZulipBinding;
   }
 
+  // Stored user agent header, since it remains constant.
+  Map<String, String>? _userAgentHeader;
+
+  @override
+  Future<BaseDeviceInfo?> get deviceInfo => _deviceInfo;
+  late Future<BaseDeviceInfo?> _deviceInfo;
+
+  @override
+  BaseDeviceInfo? get maybeDeviceInfo => _maybeDeviceInfo;
+  BaseDeviceInfo? _maybeDeviceInfo;
+
+  @override
+  Future<PackageInfo?> get packageInfo => _packageInfo;
+  late Future<PackageInfo?> _packageInfo;
+
+  @override
+  PackageInfo? get maybePackageInfo => _maybePackageInfo;
+  PackageInfo? _maybePackageInfo;
+
+  Future<BaseDeviceInfo?> _prefetchDeviceInfo() async {
+    try {
+      final info = await device_info_plus.DeviceInfoPlugin().deviceInfo;
+      _maybeDeviceInfo = switch (info) {
+        device_info_plus.AndroidDeviceInfo() => AndroidDeviceInfo(release: info.version.release,
+                                                                  sdkInt: info.version.sdkInt),
+        device_info_plus.IosDeviceInfo()     => IosDeviceInfo(systemVersion: info.systemVersion),
+        device_info_plus.MacOsDeviceInfo()   => MacOsDeviceInfo(majorVersion: info.majorVersion,
+                                                                minorVersion: info.minorVersion,
+                                                                patchVersion: info.patchVersion),
+        device_info_plus.WindowsDeviceInfo() => WindowsDeviceInfo(),
+        device_info_plus.LinuxDeviceInfo()   => LinuxDeviceInfo(name: info.name,
+                                                                versionId: info.versionId),
+        _                                    => throw UnimplementedError(),
+      };
+    } catch (e) {
+      assert(debugLog('Failed to prefetch device info: $e'));
+      // TODO(log)
```

Comment:
nit:
```suggestion
      assert(debugLog('Failed to prefetch device info: $e')); // TODO(log)
```

This style makes a grep for "TODO(log)" comments more informative, as the very line it's on is self-explanatory.